src/dist/failure_detector/failure_detector.cpp

```diff
@@ -214,8 +214,13 @@ void failure_detector::process_all_records()
                 }
             }
 
+            /*
+             * "Check interval" and "send beacon" are interleaved, so we must
+             * test if "record will expire before next time we check all the records"
+             * in order to guarantee the perfect fd
+             */
             if (record.is_alive
-                && now - record.last_send_time_for_beacon_with_ack >= _lease_milliseconds)
+                && now + _check_interval_milliseconds - record.last_send_time_for_beacon_with_ack > _lease_milliseconds)
```

Comment:
Nice job, @shengofsun. This is indeed a problem, and I did not realize this before. Thank you very much. The fix seems valid to me except that now the real lease is smaller than the specified lease, i.e., the replica server may claim to be offline even the elapsed time does not exceed the "lease".  Let's add this into the description of lease option in config.